### PR TITLE
[RFC] When importing events preserve VTIMEZONES

### LINKED ISSUE #498: DTSTART/DTEND has invalid or incomprehensible timezone information from Outlook appointment
I have the following appointment:

```
BEGIN:VCALENDAR
METHOD:REQUEST
PRODID:Microsoft Exchange Server 2010
VERSION:2.0
BEGIN:VTIMEZONE
TZID:Pacific Standard Time
BEGIN:STANDARD
DTSTART:16010101T020000
TZOFFSETFROM:-0700
TZOFFSETTO:-0800
RRULE:FREQ=YEARLY;INTERVAL=1;BYDAY=1SU;BYMONTH=11
END:STANDARD
BEGIN:DAYLIGHT
DTSTART:16010101T020000
TZOFFSETFROM:-0800
TZOFFSETTO:-0700
RRULE:FREQ=YEARLY;INTERVAL=1;BYDAY=2SU;BYMONTH=3
END:DAYLIGHT
END:VTIMEZONE
BEGIN:VEVENT
ORGANIZER:John Doe
DESCRIPTION;LANGUAGE=en-US:Wrong
UID:A15C1BDB-F417-463F-A6FA-5FBAE07805A8
SUMMARY;LANGUAGE=en-US:Broken time
DTSTART;TZID=Pacific Standard Time:20160902T123000
DTEND;TZID=Pacific Standard Time:20160902T130000
CLASS:PUBLIC
PRIORITY:5
DTSTAMP:20160901T174424Z
TRANSP:OPAQUE
STATUS:CONFIRMED
SEQUENCE:0
LOCATION;LANGUAGE=en-US:Phone (1.555.555.5555)
X-MICROSOFT-CDO-APPT-SEQUENCE:0
X-MICROSOFT-CDO-OWNERAPPTID:1234
X-MICROSOFT-CDO-BUSYSTATUS:TENTATIVE
X-MICROSOFT-CDO-INTENDEDSTATUS:BUSY
X-MICROSOFT-CDO-ALLDAYEVENT:FALSE
X-MICROSOFT-CDO-IMPORTANCE:1
X-MICROSOFT-CDO-INSTTYPE:0
X-MICROSOFT-DISALLOW-COUNTER:FALSE
BEGIN:VALARM
DESCRIPTION:REMINDER
TRIGGER;RELATED=START:-PT15M
ACTION:DISPLAY
END:VALARM
END:VEVENT
END:VCALENDAR
```

khal imports this file just fine, A-OK. It's got the right time displayed and everything.

However, as when I go to edit the description in interactive mode and *save* the appointment I get these warnings:

```
warning: DTSTART has invalid or incomprehensible timezone information in A15C1BDB-F417-463F-A6FA-5FBAE07805A8.ics in me@gmail.com
warning: DTEND has invalid or incomprehensible timezone information in A15C1BDB-F417-463F-A6FA-5FBAE07805A8.ics in me@gmail.com
```

Apparently something is wonky now that I look at it, because when I sync with my gmail calendar the time shifts to 7:30AM CDT, instead of the 2:30PM it *should* be.

Interestingly, if I edit the time of the appointment (say, change it to 2:29 PM) then save it, it gets the proper time.

Any clue what's going on here?

### PR DESCRIPTION
**Pull request by geier**
This is an important fix for an issue reported in #498 by @waynew.
The underlying issue can lead to data-loss or rather wrongly displayed events, see draft for release announcement.

@untitaker @hobarrera while this is not entirely finished yet, it would be great if you had the time to look over this already, as I want to get it out as soon as possible.

Todo
====

- [x] re-enable setting of random-uid
- [x] backport to v0.8.x
- [ ] put VEVENTs into ics file in "right" order (deferred, because of a nasty circular dependency if we were to import `sort_key` into `khal/utils.py` that needs to be resolved first, but should be okay, as we keep the original order and when actually inserting it into the db we order it correctly 

fixes #498

### CODE REVIEW

**Comment by waynew** [comment]
If there's anything that I can test out I'll be happy to :)

File: khal/utils.py [hunk 1]
```diff
@@ -590,19 +593,73 @@ def new_event(locale, dtstart=None, dtend=None, summary=None, timezone=None,
...
+
+                if not hasattr(datetime_, 'tzinfo'):
+                    continue
+
+                # check for datetimes' timezones which are not understood by
+                # icalendar
+                if datetime_.tzinfo is None and 'TZID' in item.params and \
+                        item.params['TZID'] not in missing_tz:
+                    logger.warn(
+                        'Cannot find timezone `{}` in .ics file, using default timezone. '
```
Comment (untitaker) [inline khal/utils.py]: I think khal should print the event and ask for the tz

File: khal/khalendar/utils.py [hunk 2]
```diff
@@ -133,18 +133,21 @@ def sanitize(vevent, default_timezone, href='', calendar=''):
...
     for prop in ['DTSTART', 'DTEND', 'DUE', 'RECURRENCE-ID']:
         if prop in vevent and invalid_timezone(vevent[prop]):
+            timezone = vevent[prop].params.get('TZID')
             value = default_timezone.localize(vevent.pop(prop).dt)
             vevent.add(prop, value)
-            logger.warn('{} has invalid or incomprehensible timezone '
-                        'information in {} in {}'.format(prop, href, calendar))
+            logger.warn(
+                "{} localized in invalid or incomprehensible timezone `{}` in {}/{}. "
+                "This could lead to this event being wronlgy displayed."
```
Comment (untitaker) [inline khal/khalendar/utils.py]: Typo.

File: khal/utils.py [hunk 3]
```diff
@@ -37,12 +38,14 @@
 
 
 def timefstr(dtime_list, timeformat):
-    """converts a time (as a string) to a datetimeobject
```
Comment (untitaker) [inline khal/utils.py]: Unrelated change, also why is that API designed like that? Why not take just a time as a string? Why does it delete the item from the list?

File: khal/utils.py [hunk 4]
```diff
@@ -590,19 +593,73 @@ def new_event(locale, dtstart=None, dtend=None, summary=None, timezone=None,
...
 
-    :param random_uid: asign the same random UID to all events
-    :type random_uid: bool
+    and sort the right VTIMEZONEs accordingly
+    ignores all other ics components
+    :type ics: str
+    :rtype list:
+    """
+    cal = icalendar.Calendar.from_ical(ics)
+    vevents = [item for item in cal.walk() if item.name == 'VEVENT']
```
Comment (untitaker) [inline khal/utils.py]: Unnecessary list. Can be again done in a single loop.

File: khal/controllers.py [hunk 5]
```diff
@@ -528,27 +528,28 @@ def import_ics(collection, conf, ics, batch=False, random_uid=False, format=None
...
 
 def import_event(vevent, collection, locale, batch, random_uid, format=None, env=None):
-    """import one event into collection, let user choose the collection"""
+    """import one event into collection, let user choose the collection
 
+    :type vevent: list of vevents, which can be more than one VEVENT, i.e., the
+        same UID, i.e., one "master" event and (optionally) 1+ RECURRENCE-ID events
+    :type vevent: list(str)
+    """
+    # TODO re-enable random_uid
```
Comment (geier) [inline khal/controllers.py]: calling `khal import --random-uid somefile.ics` would import all events but assign a random uid to each VEVENT, I want to re-enable that functionality.

File: khal/utils.py (hunk 3 above)
Comment (geier) [inline khal/utils.py]: for consistency reason with the other `xfstr()`s

File: khal/utils.py (hunk 1 above)
Comment (geier) [inline khal/utils.py]: probably a good idea, not sure if we need to address it with this PR though.

File: khal/controllers.py [hunk 6]
```diff
@@ -528,27 +528,28 @@ def import_ics(collection, conf, ics, batch=False, random_uid=False, format=None
...
     # print all sub-events
-    for sub_event in vevent:
+    cal = icalendar.Calendar.from_ical(vevent)
+    for sub_event in [item for item in cal.walk() if item.name == 'VEVENT']:
```
Comment (waynew) [inline khal/controllers.py]: Using `()` instead of `[]` would at least make it a generator comprehension instead of listcomp, and have the same effect of skipping creation of an intermediate list.

Comment (untitaker) [inline khal/controllers.py]: There's still overhead. I don't see a reason to use either list comprs or generator exprs.

File: khal/utils.py (hunk 1 above)
Comment (hobarrera) [inline khal/utils.py]: It might be useful as a separate command, like `repair`, which might eventually have other features as well.

I agree that it's best left outside the scope of this issue (so we can close this bug faster).

**Comment by geier** [comment]
@pimutils/packaging There is a new release of khal (0.8.4), as it fixes a bug in `import` that can lead to data loss, please update all packages as soon as possible. If you need any help with that, please ping me.
 Also, if you are maintaining a v0.7.x package that can not be upgrade because of the python 2/3 issue (and people are actually using it), ping me, so I can backport this fix onto 0.7.0.

**Comment by part1zano** [comment]
Updating the port right now. I guess it might get committed during the following week or so.

**Comment by part1zano** [comment]
Couldn't upgrade the port now as it contains py3-specific code. Could you kindly release something like 0.7.5?

**Comment by untitaker** [comment]
Khal 0.8 doesn't support Python 2.

**Comment by untitaker** [comment]
Ok, it appears @geier offered to backport this to 0.7 anyway, so nvm

**Comment by part1zano** [comment]
@untitaker thank you for the info!

**Comment by untitaker** [comment]
The same thing applies for vdirsyncer 0.13 btw, but it has been outputting warnings about Python 2 for a long time now.

**Comment by untitaker** [comment]
BTW feel free to open issues if you want to have a different way to be notified about releases.

**Comment by mathstuf** [comment]
Building for Rawhide now; will copy down to stable releases afterwards.

As for notifications, a separate issue with a better subject line would be better at least.

**Comment by mathstuf** [comment]
Fedora 23 is still using the Python 2 version, so a new 0.7 would be good there.

**Comment by part1zano** [comment]
Regarding notifications, I wish a system that notifies me via email. The mentions thing implemented here works just fine in this case. If the separate issue approach mentioned above works the same way, I'm for it.

**Comment by untitaker** [comment]
Let's continue in https://github.com/pimutils/meta/issues/5
